recipe/meta.yaml

```
@@ -36,128 +43,142 @@ source:
+        - git       # [not win]
+        - m2-git    # [win]
+        - libgomp   # [linux]
+        - llvm-openmp    # [osx]
+        - cmake
+        - ninja
+        # Keep libprotobuf here so that a compatibile version
+        # of protobuf is installed between build and host
+        - libprotobuf
+        - protobuf
+        - make      # [linux]
+      host:
+        # GPU requirements
+        - cudnn                           # [cuda_compiler_version != "None"]
+        - nccl                            # [cuda_compiler_version != "None"]
+        - magma                           # [cuda_compiler_version != "None"]
+        - cuda-version {{ cuda_compiler_version }}  # [cuda_compiler_version != "None"]
+        {% if cuda_major >= 12 %}
+        - cuda-driver-dev
+        - cuda-cudart-dev
+        - cuda-nvrtc-dev
+        - cuda-nvtx-dev
+        - cuda-nvml-dev
+        - cuda-profiler-api
+        - libcublas-dev
+        - libcufft-dev
+        - libcurand-dev
+        - libcusolver-dev
+        - libcusparse-dev
+        {% endif %}
+        # other requirements
+        - python 3.12  # [megabuild]
+        - python       # [not megabuild]
+        - numpy *      # [megabuild]
+        - numpy        # [not megabuild]
+        - pip
+        - setuptools
+        - pyyaml
+        - requests
+        - future
+        - six
+        - mkl-devel {{ mkl }}   # [blas_impl == "mkl"]
+        - libcblas * *_mkl      # [blas_impl == "mkl"]
+        - libcblas              # [blas_impl != "mkl"]
+        - liblapack             # [blas_impl != "mkl"]
+        - libgomp   # [linux]
+        - llvm-openmp    # [osx]
+        - libabseil
+        - libprotobuf
+        - sleef
+        - typing
+        - libuv
+        - pkg-config  # [unix]
+        - typing_extensions
+      run:
+        # GPU requirements without run_exports
+        - {{ pin_compatible('cudnn') }}                       # [cuda_compiler_version != "None"]
+      run_constrained:
+        # These constraints ensure conflict between pytorch and
+        # pytorch-cpu 1.1 which we built before conda-forge had GPU infrastructure
+        # built into place.
+        # https://github.com/conda-forge/pytorch-cpu-feedstock/issues/65
+        - pytorch-cpu =={{ version }}  # [cuda_compiler_version == "None"]
+        - pytorch-gpu ==99999999       # [cuda_compiler_version == "None"]
+        - pytorch-gpu =={{ version }}  # [cuda_compiler_version != "None"]
+        - pytorch-cpu ==99999999       # [cuda_compiler_version != "None"]
+        - pytorch {{ version }} cuda{{ cuda_compiler_version | replace('.', '') }}_*_{{ PKG_BUILDNUM }}  # [cuda_compiler_version != "None"]
+        - pytorch {{ version }} cpu_{{ blas_impl }}_*_{{ PKG_BUILDNUM }}                                 # [cuda_compiler_version == "None"]
+
+    test:
+      commands:
+        # libraries
+        {% for each_lib in libs %}
+        - test -f $PREFIX/lib/{{ each_lib }}${SHLIB_EXT}
+        {% endfor %}
```

```suggestion
        # libraries
        {% for each_lib in [ 'libc10', 'libshm', 'libtorch', 'libtorch_cpu', 'libtorch_global_deps'] %}
        - echo testing for the presence $PREFIX/lib/{{ each_lib }}${SHLIB_EXT}   # [unix]
        - test -f $PREFIX/lib/{{ each_lib }}${SHLIB_EXT}   # [unix]
        {% endfor %}
        {% for each_lib in ['libc10_cuda', 'libcaffe2_nvrtc', 'libtorch_cuda', 'libtorch_cuda_linalg'] %}
        - echo testing for the presence of $PREFIX/lib/{{ each_lib }}${SHLIB_EXT}  # [unix and cuda_compiler_version != "None"]
        - test -f $PREFIX/lib/{{ each_lib }}${SHLIB_EXT}  # [unix and cuda_compiler_version != "None"]
        {% endfor %}
```

In this case, I don't think the added indirection of having the libraries listed at the top "saves" any cognitive load.

It would be good to list of libraries being tested next to the test itself.
The echo command has also helped me troubleshoot recipes since the failed `test -f` doesn't often print out what file was missing.

---

> The echo command has also helped me troubleshoot recipes since the failed `test -f` doesn't often print out what file was missing.

I've never seen that happen... The failing line (with all the substitutions done) gets printed in the logs. I've left out the echos for now, we can add them if we do ever run into a problem with that.

For reference, here's the log from the last run, i.e. without any echos:
```
export PREFIX=/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1716077967897/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p
export SRC_DIR=/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1716077967897/test_tmp
+ test -f $PREFIX/lib/libc10.so
+ test -f $PREFIX/lib/libshm.so
+ test -f $PREFIX/lib/libtorch.so
+ test -f $PREFIX/lib/libtorch_cpu.so
+ test -f $PREFIX/lib/libtorch_global_deps.so
+ test -f $PREFIX/lib/libc10_cuda.so
+ test -f $PREFIX/lib/libcaffe2_nvrtc.so
+ test -f $PREFIX/lib/libtorch_cuda.so
+ test -f $PREFIX/lib/libtorch_cuda_linalg.so
+ exit 0
```